### PR TITLE
Optimize predictors for multipass motion estimation

### PR DESCRIPTION
**Pull request by KyleSiefring**
Since we perform multiple passes during motion estimation, it makes
sense to sample the bottom and right edges for predictors.
|    PSNR | PSNR Cb | PSNR Cr | PSNR HVS |    SSIM | MS SSIM | CIEDE 2000 |
|    ---: |    ---: |    ---: |     ---: |    ---: |    ---: |       ---: |
| -0.3470 | -0.4163 | -0.4998 |  -0.3425 | -0.3951 | -0.3994 |    -0.4004 |

https://beta.arewecompressedyet.com/?job=master-b17d5009f7e539aba2e1abde9cbb5203f1aca3f3&job=me_port_improved_predictors2%402020-08-06T17%3A44%3A58.052Z

### CODE REVIEW

File: src/me.rs [hunk 1]
```diff
@@ -90,13 +90,13 @@ pub fn get_subset_predictors<T: Pixel>(
   tile_bo: TileBlockOffset, cmvs: ArrayVec<[MotionVector; 7]>,
   tile_mvs: &TileMotionVectors<'_>, frame_ref_opt: Option<&ReferenceFrame<T>>,
   ref_frame_id: usize, bsize: BlockSize,
-) -> ArrayVec<[MotionVector; 16]> {
-  let mut predictors = ArrayVec::<[_; 16]>::new();
+) -> ArrayVec<[MotionVector; 17]> {
+  let mut predictors = ArrayVec::<[_; 17]>::new();
```
Comment (lu-zero) [inline src/me.rs:94]: If you are going to change it often make it a type so you would not have to keep in mind to change it in two places.

File: src/me.rs [hunk 2]
```diff
@@ -112,26 +112,49 @@ pub fn get_subset_predictors<T: Pixel>(
     add_cand(&mut predictors, mv);
   }
 
-  // EPZS subset A and B predictors.
```
Comment (ycho) [inline src/me.rs:115]: Any reason you didn't want to keep this line of EPZS comment?

**Comment by tdaede** [comment]
The only thing I don't like about this is that it removes the distinction of which predictors are from a previous pass and which are from fully RDO'd neighbor blocks, because both are stored in tile_mvs.

**Comment by ycho** [comment]
> The only thing I don't like about this is that it removes the distinction of which predictors are from a previous pass and which are from fully RDO'd neighbor blocks, because both are stored in tile_mvs.

Yeap I understand. I think it can only be knowable at best from implicit information, unless there is separate storage or indicator flag.

File: src/me.rs [hunk 3]
```diff
@@ -112,26 +112,49 @@ pub fn get_subset_predictors<T: Pixel>(
...
 
+  // Sample the center of the current block.
+  add_cand(
+    &mut predictors,
+    tile_mvs[tile_bo.0.y + clipped_half_h][tile_bo.0.x + clipped_half_w],
+  );
+
+  // Sample the middle of all blocks bordering this one.
+  // Note: If motion vectors haven't been precomputed to a given blocksize, then
+  // then the right and bottom edges will be duplicates of the center predictor
```
Comment (ycho) [inline src/me.rs:128]: 'then' repeated.

**Pull request merged**